Update OpenPedCan-analysis README.md and CONTRIBUTING.md

logstar opened the pull request:
<!--Hi there, thanks for your contribution! Please take a moment to fill out this template to facilitate the review of your pull request.-->

### Purpose/implementation Section

Update OpenPedCan-analysis README.md and CONTRIBUTING.md to update/add instructions on Docker image usage, pull request reviews, and coordinated merging of approved pull requests.

#### What was your approach?

Updated README.md Docker Image section to add:

- Instructions on how to edit Dockerfile to add dependencies.
- Explanations on why it is efficient to use the project Docker image.

Updated CONTRIBUTING.md:

- Changed Open-PBTA to OpenPedCan.
- Changed master branch to dev branch.
- Added instructions for pull request reviews.
- Add instructions on how to coordinate pull request merging.

#### What GitHub issue does your pull request address?

<https://github.com/PediatricOpenTargets/ticket-tracker/issues/28>

### Directions for reviewers. Tell potential reviewers what kind of feedback you are soliciting.

#### Which areas should receive a particularly close look?

Pull request review instructions in CONTRIBUTING.md.

Changes in the Docker Image section of README.md.

#### Is there anything that you want to discuss further?

The README.md still has a lot of OpenPBTA references, which may be confusing to the readers. I was wondering if the README.md should be updated to remove all OpenPBTA references and only explain that the repository is a fork from the OpenPBTA. This way, we may also be able to convert the OpenPedCan-analysis into a standalone repository and move the forks of OpenPedCan-analysis into a new network, by [contacting GitHub support](https://docs.github.com/en/github/setting-up-and-managing-your-github-profile/managing-contribution-graphs-on-your-profile/why-are-my-contributions-not-showing-up-on-my-profile#commit-was-made-in-a-fork).

I was also wondering if we could add a summary of the OpenPedCan-analysis at the beginning of the README.md.

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?

Yes

jharenza commented on CONTRIBUTING.md:
```diff
@@ -19,31 +19,31 @@ _If you are new to git and GitHub, check out [this Hello World guide](https://gu
-You do not want to commit changes to your `master` branch, any changes to your `master` branch should come from the upstream `master` branch.
+You do not want to commit changes to your `dev` branch, any changes to your `dev` branch should come from the upstream `dev` branch.
 
 ### Size and composition of pull requests
```
Can you also add a recommended size of new PR as 400 lines of code, and if the PR will exceed, to break it down into smaller PRs - can refer to "stacked PRs" below.

logstar replied:
Sure. Will do.

jharenza commented on CONTRIBUTING.md:
```diff
@@ -153,10 +153,25 @@ It is important to note that, for authorship, these should be substantial intell
-Participants in this project should review pull requests, which can be done using [GitHub's review interface](https://help.github.com/articles/about-pull-request-reviews/ "GitHub: about pull request reviews").
-They should suggest modifications or, potentially, directly edit the pull request to make suggested changes.
-As a reviewer, it's helpful to note the type of review you performed: did you look over the source code, did you run the source code, did you look at and interpret the results or a combination of these?
+All pull requests will undergo peer review. The review intends to ensure that a pull request conforms to the following basic quirements.
```
```suggestion
All pull requests will undergo peer review. The review intends to ensure that a pull request conforms to the following basic requirements.
```

logstar replied:
Thank you for catching the typo!

jharenza commented on CONTRIBUTING.md:
```diff
@@ -153,10 +153,25 @@ It is important to note that, for authorship, these should be substantial intell
+Participants in this project should review pull requests, which can be done using [GitHub's review interface](https://help.github.com/articles/about-pull-request-reviews/ "GitHub: about pull request reviews"). Following are the basic guidelines for reviewing pull requests:
+
+- Note the type of review you performed: did you look over the source code, did you look over the documentation, did you run the source code, did you look at and interpret the results or a combination of these?
+- Suggest modifications or, potentially, directly edit the pull request to make suggested changes, so that the pull request conforms to the aforementioned basic requirements.
```
```suggestion
- Suggest modifications or, potentially, directly suggest, _but do not commit changes to_, the pull request, so that the pull request conforms to the aforementioned basic requirements.
```

jharenza commented on CONTRIBUTING.md:
```diff
@@ -153,10 +153,25 @@ It is important to note that, for authorship, these should be substantial intell
+
+- Note the type of review you performed: did you look over the source code, did you look over the documentation, did you run the source code, did you look at and interpret the results or a combination of these?
+- Suggest modifications or, potentially, directly edit the pull request to make suggested changes, so that the pull request conforms to the aforementioned basic requirements.
+- Explain in **details** why the suggested modifications are **necessary**, and how the modifications should be implemented specifically, so that the developers are able to follow the suggestions without misunderstanding.
```
```suggestion
- Explain in **detail** why the suggested modifications are **necessary**, and how the modifications should be implemented specifically, so that the developers are able to follow the suggestions without misunderstanding.
```

jharenza commented on CONTRIBUTING.md:
```diff
@@ -153,10 +153,25 @@ It is important to note that, for authorship, these should be substantial intell
+- Note the type of review you performed: did you look over the source code, did you look over the documentation, did you run the source code, did you look at and interpret the results or a combination of these?
+- Suggest modifications or, potentially, directly edit the pull request to make suggested changes, so that the pull request conforms to the aforementioned basic requirements.
+- Explain in **details** why the suggested modifications are **necessary**, and how the modifications should be implemented specifically, so that the developers are able to follow the suggestions without misunderstanding.
+- If the suggested modifications are extensive refactoring or re-designing without any change on the code behaviors, results or run-time, consider submitting a new issue for the refactoring or re-designing, and discussing the priority and specific implementations in the new issue, after the pull request under review is merged, so that the pull request under review will not be kept open for length ...
```
👏

jharenza commented on README.md:
```diff
@@ -286,8 +286,7 @@ Files that are intermediate, which means that they are useful within an analysis
 
-To add dependencies that are required for your analysis to the project Docker image, you must alter the project [`Dockerfile`](https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/master/Dockerfile).
-
+To add dependencies that are required for your analysis to the project Docker image, you must alter the project [`Dockerfile`](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/dev/Dockerfile). The `Dockerfile` can be directly edited to install dependencies, if you are developing using a branch on the [PediatricOpenTargets/OpenPedCan-analysis](https://github.com/PediatricOpenTarget ...
```
just a general comment that we generally prefer each sentence on a new line for ease of review :)

logstar replied:
Thank you for the comment. Good to know! Will revise accordingly.

logstar replied:
Revised in commit <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/73/commits/02e42b4bd78a4609174a88866d99a7adaf0891e1>

jharenza commented on README.md:
````diff
@@ -308,6 +308,12 @@ The most recent version of the project Docker image, which is pushed to Docker H
 docker pull pgc-images.sbgenomics.com/d3b-bixu/open-pedcan:latest
 ```
 
+Development using the project Docker image is highly recommended. An analysis that is developed using the project Docker image can be efficiently rerun by another developer or the original developer (after a long time since it is developed), without dependency or numerical issues. This would significantly facilitate the following tasks that are constantly performed by all developers of the OpenPe ...
````
```suggestion
Development should utilize the project Docker image. 
An analysis that is developed using the project Docker image can be efficiently rerun by another developer or the original developer (after a long time since it is developed), without dependency or numerical issues. 
This will significantly facilitate the following tasks that are constantly performed by all developers of the OpenPedCan-analysis project.
```

logstar replied:
Should we remove the "Local Development" section, as we require using Docker image to develop?

jharenza replied:
we can perhaps leave it for some edge cases - and we may need to add some language later about development of modules for CAVATICA, but I don't think we have the rules yet.

logstar replied:
> we can perhaps leave it for some edge cases - and we may need to add some language later about development of modules for CAVATICA, but I don't think we have the rules yet.

Sounds good. I will leave the "Local Development" section there.

jharenza reviewed (changes requested):
Hi @logstar - this looks good in general - had a few suggestions. Also, can you add a label usage section for PRs? "ready for review", "work in progress", and "merge next" should be used systematically.

logstar commented:
> Hi @logstar - this looks good in general - had a few suggestions. Also, can you add a label usage section for PRs? "ready for review", "work in progress", and "merge next" should be used systematically.

Thank you for the suggestions. I will add a label usage section for PRs and issues.

runjin326 commented on CONTRIBUTING.md:
```diff
@@ -165,7 +165,7 @@ All pull requests will undergo peer review. The review intends to ensure that a
 Participants in this project should review pull requests, which can be done using [GitHub's review interface](https://help.github.com/articles/about-pull-request-reviews/ "GitHub: about pull request reviews"). Following are the basic guidelines for reviewing pull requests:
```
I couldn't figure out how to comment on codes that are not changed but for the staggered PR part, maybe we should specify that we should push all results from the relevant scripts as well (if they are not in .gitignore)

logstar replied:
I agree that all results should also be pushed, if they are not in .gitignore.

I was wondering if you could clarify the case where you need to comment on codes that are not changed. Maybe add a paragraph as suggestion?

runjin326 replied:
Oh sorry I meant for Github when we are reviewing the PR. I couldn't find a place to comment on lines that are not changed. For example, I wanted to comment on lines for the staggered PR but since they are not changed here, I couldn't figure out how to add comments specific to those lines - and hence I am adding comments in a less relevant place. That's what I meant haha

logstar replied:
No worries. Thank you for the clarification @runjin326 . I think I get it.

I have not encountered cases like this in my work yet, so I am not sure how to best handle these cases. 

Therefore, I was wondering if you could draft a paragraph as suggestion for the best practice of handling such cases.

logstar replied:
Resolved with commit <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/73/commits/f4932a805de9f597523509569f8318860f6b6fc7>

runjin326 commented on CONTRIBUTING.md:
```diff
@@ -19,31 +19,31 @@ _If you are new to git and GitHub, check out [this Hello World guide](https://gu
 You will then be able to file a pull request from this branch: https://help.github.com/en/articles/creating-a-pull-request-from-a-fork
-4. Once your pull request is approved and merged by the project maintainers, it's time to keep the `master` branch of your fork up-to-date with the `master` branch of this repository by following this process: https://help.github.com/en/articles/syncing-a-fork.
+4. Once your pull request is approved and merged by the project maintainers, it's time to keep the `dev` branch of your fork up-to-date with the `dev` branch of this repository by following this process: https://help.github.com/en/articles/syncing-a-fork.
 Here's a typical series of steps for performing that:
```
Maybe we can add step 5 saying that the authors for the PRs should delete the remote branch that had been merged

logstar replied:
I agree. Could you add a suggestion for this purpose?

runjin326 replied:
Sure! Let me know whether this seems correct. 

```suggestion
5. Once your pull request is merged by the project maintainer, please go to the repo, find the branch that the PR is based on, and delete the branch from the repo. 
Here's a typical series of steps for performing that:
```

logstar replied:
Thank you @runjin326 for the clear description! Will commit shortly.

runjin326 commented on CONTRIBUTING.md:
````diff
@@ -72,7 +72,7 @@ We're creating the new branch with all the changes from `new-analysis` unstaged.
 ```sh
 git checkout -b new-analysis-first new-analysis
-git reset master
+git reset dev
````
I guess I have a question for my own clarification - do we want each staggered PR to only have the modifications of the codes in that particular PR? If that is the case, whenever we do modifications on the PR1, and if the results from PR1 would be carried over to PR2, how do we make sure the results are carried over? I guess a general guideline for how to do that would be helpful here as well.

logstar replied:
I think we want each staggered PR to have modifications in its prior staggered PRs, e.g. staggered PR part 3 should contain modifications in staggered PR part 1 and 2. Otherwise, a later part staggered PR may not be able to merge into the dev branch.

Ideally, PRs are evaluated and merged sequentially, i.e. PR 3 is not reviewed if PR 1 and 2 are not merged, as one major purpose of having staggered PRs is to isolate reviews for efficiency. This way, merging modifications may not be too complicated.

Practically, I think the reviewers and developers may need to work together to complete the review of multiple staggered PRs, so merging modifications may be carefully evaluated on a case-by-cases basis.

logstar replied:
I can imagine there are cases PR 3 requires changes in PR 1 and PR 2, but I think it is more efficient to handle it in PR 3 rather than leaving PR 1 and 2 open for changes. Any thoughts? @runjin326

runjin326 replied:
Gotcha! Thanks for the clarification. It makes sense!

runjin326 replied:
Oh yes didn't see your other comment - yes I think in those cases when we realize in PR3 that we need to change PR1 and PR2, we can handle that in PR3 - instead of changing from PR1 and then carry all the way over to PR3 again (while leaving PR1 and PR2 open)

runjin326 commented on README.md:
```diff
@@ -308,6 +308,14 @@ The most recent version of the project Docker image, which is pushed to Docker H
+- Review another developer's pull request, including code and results. For more information about pull request and review, see [the guideline for how to contribute to the OpenPedCan-analysis repository](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/dev/CONTRIBUTING.md#contribution-guidelines-for-the-openpbta-analysis).
+- Update the results of an analysis module that is developed by another developer. For example, rerun the same analysis module with new data.
+- Update the code of an analysis module that is developed by another developer. For example, add a new feature to a module, or refactor a module.
+
```
One comment that is from line 66-120 - we now have a lot more samples, do we want to update that part as well?

logstar replied:
I think the samples part is out of the scope of this PR and addressed issue <https://github.com/PediatricOpenTargets/ticket-tracker/issues/28>, but we could cram the update in to reduce the work of submitting a new issue and PR for that purpose.

@jharenza I was wondering if you have any preference.

runjin326 reviewed (commented):
Just a couple comments :)

runjin326 commented on CONTRIBUTING.md:
```diff
@@ -153,10 +154,25 @@ It is important to note that, for authorship, these should be substantial intell
+
+- Note the type of review you performed: did you look over the source code, did you look over the documentation, did you run the source code, did you look at and interpret the results or a combination of these?
+- Suggest modifications or, potentially, directly suggest, _but do not commit changes to_, the pull request, so that the pull request conforms to the aforementioned basic requirements.
+- Explain in **detail** why the suggested modifications are **necessary**, and how the modifications should be implemented specifically, so that the developers are able to follow the suggestions without misunderstanding.
```
Oh I get it now - how about the following:
```suggestion
- Explain in **detail** why the suggested modifications are **necessary**, and how the modifications should be implemented specifically, so that the developers are able to follow the suggestions without misunderstanding.
- For any suggestions/comments that are related to the lines that were changed in the particular PR, make the comment underneath the code by clicking the "+" next to the code. Click "Start a review" for the first comment, and afterwards, click "Add single comment" to keep all comments within the same review. Click "Finish your review" once completed. 
- If there is any comments/suggestions that are **outside** of the code changes this particular PR, for example, any questions about the results, or any lines of code (although not changed) that seemed puzzling, mentioned them as comments after you click "Finish your review". 
```

logstar replied:
Thank you for the clear descriptions @runjin326 ! Sorry that I thought what you described at <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/73#discussion_r681241142> was in a more specific context.

I think maybe we could also use permalinks for this purpose, which is included in the following suggestion. I was wondering if it makes sense.

```suggestion
- Explain in **detail** why the suggested modifications are **necessary**, and how the modifications should be implemented specifically, so that the developers are able to follow the suggestions without misunderstanding.
- For any suggestions/comments that are related to the lines that were changed in the particular PR, make the comment underneath the code by clicking the "+" next to the code. Click "Start a review" for the first comment, and afterwards, click "Add single comment" to keep all comments within the same review. Click "Finish your review" once completed. 
- If there is any comments/suggestions that are **outside** of the code changes this particular PR, for example, any questions about the results, or any lines of code (although not changed) that seemed puzzling, mentioned them as comments after you click "Finish your review". The permalinks of the referenced results or code can be obtained by the following procedure: click the "Commits" tab of the pull request, browse the repository at a specific commit that contains the referenced results or code by clicking the `<>` button of the commit row, go to the referenced file in the repository to get file or line permalinks.
```

runjin326 replied:
Hey Yuanchao @logstar, yes using the permalink would be a great idea!

logstar commented:
Hi @jharenza . 

I have updated the CONTRIBUTING.md:

- Added the recommended size of PR and clarified where to submit issues, in commit <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/73/commits/44ff04634340dd8472fdaac723d64ee344945237>.
- Added a "GitHub labels for issues and pull requests" section, in commit <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/73/commits/1ba82fc63e516193d0eb1a4a70b44f8014e8e99e>.

jharenza reviewed (approved):
thank you, looks good to me!

logstar commented:
Thank you for the reviews @jharenza @runjin326 ! I will merge this PR shortly.
